crypto/ec/ec_key.c

```diff
@@ -1112,19 +1134,67 @@ static int ecdsa_keygen_pairwise_test(EC_KEY *eckey, OSSL_CALLBACK *cb,
     OSSL_SELF_TEST_onbegin(st, OSSL_SELF_TEST_TYPE_PCT,
                            OSSL_SELF_TEST_DESC_PCT_ECDSA);
 
-    sig = ECDSA_do_sign(dgst, dgst_len, eckey);
-    if (sig == NULL)
+    group = EC_KEY_get0_group(eckey);
+    curve_name = OPENSSL_strdup(EC_curve_nid2nist(EC_GROUP_get_curve_name(group)));
+    keylen = EC_KEY_key2buf(eckey, POINT_CONVERSION_UNCOMPRESSED, &keyoct, bnctx);
+    if (keylen == 0)
+        goto err;
+    privlen = EC_KEY_priv2buf(eckey, &privoct);
+    if (privlen == 0)
+        goto err;
+    paramskey[0] = OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_GROUP_NAME, curve_name, 0);
+    paramskey[1] = OSSL_PARAM_construct_octet_string(OSSL_PKEY_PARAM_PUB_KEY, keyoct, keylen);
+    paramskey[2] = OSSL_PARAM_construct_BN(OSSL_PKEY_PARAM_PRIV_KEY, privoct, privlen);
+    paramskey[3] = OSSL_PARAM_construct_end();
+    /*
+     * Allow generating signature for a P-192 key construction, which
+     * will return unapproved (fips-approved=0) indicator
+     */
+    paramsinit[0] = OSSL_PARAM_construct_int(OSSL_SIGNATURE_PARAM_FIPS_KEY_CHECK, &sign_key_check);
+    paramsinit[1] = OSSL_PARAM_construct_end();
+    /*
+     * EC_KEY_priv2buf + construct_octet_string which allocate and set
+     * buffers, but appear to use incorrect encoding which result in
+     * signature verification failures. Hence call set_BN to set BN a
+     * second time, ensuring private key is set using correct
+     * encoding.
+     */
+    if (!OSSL_PARAM_set_BN(&paramskey[2], EC_KEY_get0_private_key(eckey)))
+        goto err;
+    fromctx = EVP_PKEY_CTX_new_from_name(libctx, "EC", propq);
+    if (fromctx == NULL)
+        goto err;
+    if (EVP_PKEY_fromdata_init(fromctx) <= 0
+            || EVP_PKEY_fromdata(fromctx, &pkey, EVP_PKEY_KEYPAIR, paramskey) <= 0)
```

Comment:
What is the motivation for rebuilding the key using `fromdata`? Why not just use the given key???